CMakeLists.txt

```diff
@@ -512,6 +512,41 @@ if(ENABLE_HARDENING)
     if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
       try_append_linker_flag("-Wl,-fixup_chains" TARGET hardening_interface)
     endif()
+
+    if(HAVE_LIBCPP)
+      # https://libcxx.llvm.org/Hardening.html
+      if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+        target_compile_options(hardening_interface INTERFACE
+          -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS_IN_STRING
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS_IN_VECTOR
+          -D_LIBCPP_ABI_BOUNDED_UNIQUE_PTR
+          -D_LIBCPP_ABI_BOUNDED_ITERATORS_IN_STD_ARRAY
```

Comment:
Not manually, but I assume CI will fail if that happens. If it fails, then I will remove it.

Reply:
Do you think it would be better to omit those `_LIBCPP_ABI_BOUNDED*`?

Reply:
I removed the `_LIBCPP_ABI_BOUNDED*` options from CMake when `ENABLE_HARDENING=ON` because they caused linking failures for the multiprocess task because libmultiprocess was compiled without them, i.e. ABI incompatibility issues. So, don't enable them on users when `ENABLE_HARDENING=ON`.

> It would be good to have them in a CI task

Right, I changed a few CI tasks to use them, enabling them manually, outside of what CMake does by default, see the second commit and the PR description.

Reply:
No, wait, it was the tsan task, not the multiprocess task! Edited the above comment of mine. Anyway, it does not matter that much - it is an ABI incompatibility, so the conclusion is the same: don't enable `_LIBCPP_ABI_BOUNDED*` on users when `ENABLE_HARDENING=ON`.

> Why would the other tasks not break the ABI?

I am not sure exactly, but it has probably something to do with the way libc++ is compiled or `_LIBCPP_HARDENING_MODE_FAST` not playing well with `_LIBCPP_ABI_BOUNDED*`. Here is the exact failure: https://api.cirrus-ci.com/v1/task/5304194619408384/logs/ci.log, see `undefined reference to ... std::__1::basic_string<...>::insert(std::__1::__bounded_iter<...`. This error just does not happen on the other tasks.

> Just because CI is green doesn't mean the change is correct.

If there are ABI incompatibilities, then it will fail to compile with undefined reference because it cannot find `__bounded_iter` or `std::__debug::whatever` in the standard library.